src/instance.rs

```diff
@@ -71,55 +209,82 @@ where
     T: PyTypeInfo,
 {
     /// Borrows a GIL-bound reference to the contained `T`. By binding to the GIL lifetime, this
-    /// allows the GIL-bound reference to not require `Python` for any of its methods.
+    /// allows the GIL-bound reference to not require [`Python<'py>`](crate::Python) for any of its methods,
+    /// which makes calling methods on it more ergonomic.
     ///
     /// For native types, this reference is `&T`. For pyclasses, this is `&PyCell<T>`.
     ///
+    /// Note that the lifetime of the returned reference is the shortest of `&self` and
+    /// [`Python<'py>`](crate::Python).
+    /// Consider using [`Py::into_ref`] instead if this poses a problem.
+    ///
     /// # Examples
+    ///
     /// Get access to `&PyList` from `Py<PyList>`:
     ///
     /// ```
     /// # use pyo3::prelude::*;
     /// # use pyo3::types::PyList;
-    /// # Python::with_gil(|py| {
-    /// let list: Py<PyList> = PyList::empty(py).into();
-    /// let list: &PyList = list.as_ref(py);
-    /// assert_eq!(list.len(), 0);
-    /// # });
+    /// #
+    /// Python::with_gil(|py| {
+    ///     let list: Py<PyList> = PyList::empty(py).into();
+    ///     let list: &PyList = list.as_ref(py);
+    ///     assert_eq!(list.len(), 0);
+    /// });
     /// ```
     ///
     /// Get access to `&PyCell<MyClass>` from `Py<MyClass>`:
     ///
     /// ```
     /// # use pyo3::prelude::*;
+    /// #
     /// #[pyclass]
     /// struct MyClass { }
-    /// # Python::with_gil(|py| {
-    /// let my_class: Py<MyClass> = Py::new(py, MyClass { }).unwrap();
-    /// let my_class_cell: &PyCell<MyClass> = my_class.as_ref(py);
-    /// assert!(my_class_cell.try_borrow().is_ok());
-    /// # });
+    ///
+    /// Python::with_gil(|py| {
+    ///     let my_class: Py<MyClass> = Py::new(py, MyClass { }).unwrap();
+    ///     let my_class_cell: &PyCell<MyClass> = my_class.as_ref(py);
+    ///     assert!(my_class_cell.try_borrow().is_ok());
+    /// });
     /// ```
     pub fn as_ref<'py>(&'py self, _py: Python<'py>) -> &'py T::AsRefTarget {
         let any = self.as_ptr() as *const PyAny;
         unsafe { PyNativeType::unchecked_downcast(&*any) }
     }
 
-    /// Similar to [`as_ref`](#method.as_ref), and also consumes this `Py` and registers the
+    /// Similar to [`as_ref`](#method.as_ref) but consumes `self` and registers the
```

Comment:
Should mention that this has a cost; otherwise it sounds like `into_ref` is always preferable.

Reply:
Is the additional cost just in situations where a `Py<T>` is needlessly created and destroyed? Or are there performance penalties beyond that?

Reply:
It requires placing the pointer into pyO3's internal pool, so will involve a thread-local access. Tbh I'm not sure how much that's relevant right now as most of PyO3's APIs do that!